File: Juego/Invasiones/fuente/Sprites/Sprite.cs
using System;
using System.Collections.Generic;
using System.Text;
using Tao.Sdl;
using Invasiones.Debug;
using Invasiones.Dibujo;
using System.Collections;
using Invasiones.Recursos;
using System.Drawing;
using Invasiones.Map;

namespace Invasiones.Sprites
{
	/// <summary>
	/// Clase que contiene los sprites. Permite cargar, animar un sprite. Se trata de
	/// un tipo de m_mapa de bits dibujados en la pantalla de ordenador por  el hardware,
	/// sin cálculos adicionales de la CPU. Poseen transparencias, dejándoles así
	/// asumir otras formas a la del rectángulo.
	/// </summary>
	public class Sprite
	{

		/// <summary>
		/// Las animaciones que tiene el sprite.
		/// </summary>
		protected Animaciones[] m_animaciones;

		/// <summary>
		/// La animacion actual, la que esta reproduciendo en un instante dado.
		/// </summary>
		protected Animaciones m_animacionActual;

		/// <summary>
		/// El id de la animacion actual.
		/// </summary>
		private int m_idAnimacionActual;

		public Sprite()
		{
		}

		/// <summary>
		/// Constructor de copia del Sprite
		/// </summary>
		/// <param name="spr"> el Sprite que se quiere copiar.</param>
		public Sprite(ref Sprite spr)
		{
			m_animaciones = new Animaciones[spr.m_animaciones.Length];

			for (int i = 0; i < spr.m_animaciones.Length; i++)
			{
				m_animaciones[i] = new Animaciones(spr.m_animaciones[i]);
			}

			m_animacionActual = m_animaciones[0];
		}

		/// <summary>
		/// Destructor de la clase.
		/// </summary>
		~Sprite()
		{
			this.Dispose();
		}

		/// <summary>
		/// Libera cualquier recurso utilizado.
		/// </summary>
		public virtual void Dispose()
		{
			GC.SuppressFinalize(this);
		}

		/// <summary>
		/// Método abstracto a ser pisado por las clases derivadas. Actualiza el sprite.
		/// </summary>
		public void Actualizar()
		{
			m_animacionActual.Actualizar();
		}

		/// <summary>
		/// Setea true o false si la animacion loopea  o no. 
		/// </summary>
		public bool Loop
		{
			get { return m_animacionActual.Loop; }
			set { m_animacionActual.Loop = value; }
		}

		/// <summary>
		/// Setea la animación dada por parámetro en el frame 0.
		/// </summary>
		/// <param name="anim">el id de la animación</param>
		/// <returns>true si pudo cargar correctamente.</returns>
		public bool SetearAnimacion(int anim)
		{

			if (m_idAnimacionActual == anim)
			{
				return false;
			}
			m_idAnimacionActual = anim;

			int resta = 0;
			int animacionesAnteriores = 0;

			for (int i = 0; i < m_animaciones.Length; i++)
			{
				if (anim >= animacionesAnteriores && anim - animacionesAnteriores < m_animaciones[i].CantidadDeAnimaciones)
				{
					m_animacionActual = m_animaciones[i];
					resta = animacionesAnteriores;
				}
				animacionesAnteriores += m_animaciones[i].CantidadDeAnimaciones;
			}

			m_animacionActual.SetearAnimacion(anim - resta);

			return true;

		}

		/// <summary>
		/// Agrega animaciones a la clase. Es utilizada por la clase AdministradorDeUnidades
		/// </summary>
		/// <param name="i">El índice de la animación</param>
		/// <param name="anim">El objeto que contiene la información de la animación.</param>
		/// <returns>true si pudo agregar la animaciçon.</returns>
		public bool AgregarAnimacion(int i, Animaciones anim)
		{
			if (m_animaciones == null)
			{
				Log.Instancia.Error("No se carga la unidad porque no esta seteada la cantidad de animaciones que tiene el sprite.");
				return false;
			}

			if (i >= m_animaciones.Length)
			{
				Log.Instancia.Debug("La animacion qeu se quiere agregar tiene un indice inválido.");
				return false;
			}
			m_animaciones[i] = anim;
			return true;
		}


		/// <summary>
		/// Dibuja el sprite en la pantalla
		/// </summary>
		/// <param name="g">El Video en donde dibujarlo.</param>
		public void Dibujar(Video g, int x, int y)
		{
			g.Dibujar(m_animacionActual.Imagen, x, y, 0);
			//g.Dibujar(m_imagen, x, y, 0);
		}

		/// <summary>
		/// Carga la imagen asociada al sprite.
		/// </summary>
		public bool Cargar()
		{
			bool ok = true;
			for (int i = 0; i < m_animaciones.Length; i++)
			{
				if (!m_animaciones[i].Cargar())
				{
					ok = false;
				}
			}
			m_animacionActual = m_animaciones[0];

			return ok;
		}

		/// <summary>
		/// Las animaciones que contiene el Sprite.
		/// </summary>
		public Animaciones[] Animaciones
		{
			get { return m_animaciones; }
			set { m_animaciones = value; }
		}

		/// <summary>
		/// El ancho del frame.
		/// </summary>
		public int FrameAncho
		{
			get { return m_animacionActual.FrameAncho; }
		}

		/// <summary>
		/// El alto del frame.
		/// </summary>
		public int FrameAlto
		{
			get { return m_animacionActual.FrameAlto; }
		}


		public int CantidadDeFrames
		{
			get { return m_animacionActual.CantidadDeFrames; }
		}
		///<summary>
		///Devuelve la imagen del sprite.
		///</summary>
		public Superficie Imagen
		{
			get { return m_animacionActual.Imagen; }
		}

		/// <summary>
		///  Setea el sprite en estado reproduciendo.
		/// </summary>
		public void Reproducir()
		{
			m_animacionActual.Reproducir();
		}

		/// <summary>
		///  Setea el sprite en estado reproduciendo = false
		/// </summary>
		public void Parar()
		{
			m_animacionActual.Parar();
		}

		public Point Offsets
		{
			get { return m_animacionActual.Offsets; }
		}


		public bool TerminoDeAnimar()
		{
			return m_animacionActual.TerminoDeAnimar();
		}

		public int FrameActual
		{
			get
			{
				return m_animacionActual.FrameActual;
			}
		}

		public int AnimacionActual
		{
			get
			{
				return m_animacionActual.AnimacionActual;
			}
		}

		public void SetearFrame(int p)
		{
			if (m_animacionActual != null)
			{
				m_animacionActual.SetearFrame(p);
			}
		}
	}
}

File: macOS/Invasiones/Invasiones/Sprites/Sprite.swift
//
//  Sprite.swift
//  Invasiones
//
//  Created by Lucia Medina Fretes on 06.04.26.
//
//  Port of Sprite.cs — container of animations for a game sprite.
//  Manages an indexed array of Animaciones and delegates to the active animation.
//

import Foundation

class Sprite {

    // MARK: - Declarations
    var m_animaciones: [Animaciones?] = []
    private var m_animacionActual:   Animaciones?
    private var m_idAnimacionActual: Int = -1

    // MARK: - Initializeres
    init() {}

    /// Copy initializer: clones all animations from the original sprite.
    init(copia: Sprite) {
        m_animaciones    = copia.m_animaciones.map { $0.map { Animaciones(copia: $0) } }
        m_animacionActual = m_animaciones.compactMap { $0 }.first
    }

    // MARK: - Properties
    var loop: Bool {
        get { m_animacionActual?.loop ?? false }
        set { m_animacionActual?.loop = newValue }
    }

    var frameAncho:       Int          { m_animacionActual?.frameAncho          ?? 0    }
    var frameAlto:        Int          { m_animacionActual?.frameAlto           ?? 0    }
    var cantidadDeFrames: Int          { m_animacionActual?.cantidadFrames      ?? 0    }
    var imagen:           Superficie?  { m_animacionActual?.m_imagen                   }
    var offsets:          (x: Int, y: Int) { m_animacionActual?.offsets ?? (0, 0) }
    var frameActual:      Int          { m_animacionActual?.frameActual         ?? 0    }
    var animacionActual:  Int          { m_animacionActual?.animacionActual     ?? 0    }

    // MARK: - Methods

    func actualizar() {
        m_animacionActual?.actualizar()
    }

    @discardableResult
    func setearAnimacion(_ anim: Int) -> Bool {
        guard anim != m_idAnimacionActual else { return false }
        m_idAnimacionActual = anim

        var resta = 0
        var animacionesAnteriores = 0

        for animObj in m_animaciones.compactMap({ $0 }) {
            if anim >= animacionesAnteriores &&
               anim - animacionesAnteriores < animObj.cantidadAnimaciones {
                m_animacionActual = animObj
                resta = animacionesAnteriores
            }
            animacionesAnteriores += animObj.cantidadAnimaciones
        }

        m_animacionActual?.setearAnimacion(anim - resta)
        return true
    }

    @discardableResult
    func agregarAnimacion(_ i: Int, _ anim: Animaciones) -> Bool {
        guard !m_animaciones.isEmpty else {
            Log.Instancia.error("No se carga la unidad: la cantidad de animaciones no está seteada.")
            return false
        }
        guard i < m_animaciones.count else {
            Log.Instancia.debug("Animacion con indice invalido: \(i)")
            return false
        }
        m_animaciones[i] = anim
        return true
    }

    /// Pre-allocates N animation slots (equivalent to `new Animaciones[N]` in C#).
    func reservarSlots(_ count: Int) {
        m_animaciones = Array(repeating: nil, count: count)
    }

    func dibujar(_ g: Video, _ x: Int, _ y: Int) {
        guard let img = m_animacionActual?.m_imagen else { return }
        g.dibujar(img, x, y, 0)
    }

    @discardableResult
    func cargar() -> Bool {
        var ok = true
        for anim in m_animaciones.compactMap({ $0 }) {
            if !anim.cargar() { ok = false }
        }
        m_animacionActual = m_animaciones.compactMap({ $0 }).first
        return ok
    }

    func reproducir() { m_animacionActual?.reproducir() }
    func parar()      { m_animacionActual?.parar() }

    func terminoDeAnimar() -> Bool { m_animacionActual?.terminoDeAnimar() ?? false }

    func setearFrame(_ p: Int) { m_animacionActual?.setearFrame(p) }
}
